docs/devel/writing-good-e2e-tests.md

```diff
@@ -0,0 +1,168 @@
+   automates cleaning up the mess after your test has completed (it
+   just deletes everything in the namespace).  This helps to ensure
+   that tests do not leak resources.  Your tests should always use
+   this framework.  Trying other home-grown approaches to avoiding
+   name clashes and resource leaks has proven to be a very bad idea.
++
+   [E2e utils library](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/util.go):
+   This handy library provides tons of reusable code for a host of
+   commonly needed test functionality, including waiting for resources
+   to enter specified states, safely and consistently retrying failed
+   operations, usefully reporting errors, and much more.  Make sure
+   that you're familiar with what's available there, and use it.
+   Likewise, if you come across a generally useful mechanism that's
+   not yet implemented there, add it so that others can benefit from
+   your brilliance.  In particular pay attention to the variety of
+   timeout and retry related constants at the top of that file. Always
+   try to reuse these constants rather than try to dream up your own
+   values.  Even if the values there are not precisely what you would
+   like to use (timeout periods, retry counts etc), the benefit of
+   having them be consistent and centrally configurable across our
+   entire test suite typically outweighs your personal preferences.
++ **Follow the examples of stable, well-written tests:** Some of our
+   existing end-to-end tests are better written and more reliable than
+   others.  A few examples of well-written tests include:
+   [Replication Controllers](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/rc.go),
+   [Services](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/service.go),
+   [Reboot](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/reboot.go).
++ [Ginkgo Test Framework](https://github.com/onsi/ginkgo): This is the
+   test library and runner upon which our e2e tests are built.  Before
+   you write or refactor a test, read the docs and make sure that you
+   understand how it works.  In particular be aware that every test is
+   uniquely identified and described (e.g. in test reports) by the
+   concatenation of it's "Describe" clause and nested "It" clauses.
+   So for example "Describe("Pods",...).... It(""should be scheduled
+   with cpu and memory limits") produces a sane test identifier and
+   descriptor "Pods should be scheduled with cpu and memory limits",
+   which makes it clear what's being tested, and hence what's not
+   working if it fails.  Other good examples include:
+   +  "CAdvisor should be healthy on every node"
+   + "Daemon set should run and stop complex daemon"
+   > On the contrary
+(these are real examples), the following are less good test
+descriptors:
+   +  "KubeProxy should test kube-proxy"
+   + "Nodes [Disruptive] Network when a node becomes unreachable
```

Comment:
I don't want people to get confused and think the `[Disruptive]` label is what makes this a bad test name.  Either add a good name with a label in it, or find one that doesn't have a label?